src/librustc_span/source_map.rs

```diff
@@ -194,12 +209,25 @@ impl SourceMap {
         self.files.borrow().stable_id_to_source_file.get(&stable_id).map(|sf| sf.clone())
     }
 
-    fn next_start_pos(&self) -> usize {
-        match self.files.borrow().source_files.last() {
-            None => 0,
-            // Add one so there is some space between files. This lets us distinguish
-            // positions in the `SourceMap`, even in the presence of zero-length files.
-            Some(last) => last.end_pos.to_usize() + 1,
+    fn allocate_address_space(&self, size: usize) -> Result<usize, OffsetOverflowError> {
+        let size = u32::try_from(size).map_err(|_| OffsetOverflowError)?;
+
+        loop {
+            let current = self.used_address_space.load(Ordering::Relaxed);
+            let next = current
+                .checked_add(size)
+                // Add one so there is some space between files. This lets us distinguish
+                // positions in the `SourceMap`, even in the presence of zero-length files.
+                .and_then(|next| next.checked_add(1))
+                .ok_or(OffsetOverflowError)?;
+
+            if self
+                .used_address_space
+                .compare_exchange(current, next, Ordering::Relaxed, Ordering::Relaxed)
```

Comment:
I'm having a hard time grokking [the docs](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html#method.compare_exchange). Can you give me a tl;dr for this? Is this updating `used_address_space` if it already equals `current` or something? If so, then can you explain if/how `current` is changing between loop iterations? (I'm assuming(⚠️) `checked_add` isn't mutating `current`).

Reply:
> Is this updating used_address_space if it already equals current or something?

Yes. To understand why you'd want to do this, think about the use case for atomics. Typically the alternative to atomics is to use some kind of lock. In that scenario, you acquire a read/write lock on a value, perform a calculation, update the value and then release the lock. This kind of design is usually correct and easy to understand but can be bad for performance in some situations because you have to acquire a lock just to read the current value.

Atomics on the other hand, don't lock. Your CPU typically gives you special instructions or guarantees about existing instructions for certain sizes of data. These allow you to do things like read the entire value or write the entire value without the reads\writes becoming mangled with data from other threads. `compare_exchange` uses an instruction that says "if this is the same as the value I'm providing then update it". This lets you do things like the above where we read the value, perform a calculation to get a new value and if the value hasn't changed since we read it, then we can write the new value. If the value has changed, we loop around again and redo the operation until it succeeds.

This operation is sometimes called [compare and swap](https://en.wikipedia.org/wiki/Compare-and-swap).

Reply:
> if the value hasn't changed since we read it

Thank you _so_ much! This makes way more sense. I was reading this as "if the value I'm providing is the same as the value it already is, then update it" which made _no_ sense because you'd always be updating values to the values they already are :-P